### PR TITLE
Add example

### PR DESCRIPTION
**Pull request by bun919tw**
**Before submitting a pull request**, please make sure the following is done:

### Checklist

- [x] Fork [the repository](https://github.com/getamis/alice) and create your new branch from `master`.
- [x] Please mention the name of community in Pull Request title.
- [x] Git Commit Messages
    - [x] Use the present tense (Also in Pull Request title): "Add feature" not "Added feature"
    - [x] Use the imperative mood (Also in Pull Request title): "Move cursor to..." not "Moves cursor to..."
    - [x] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.

### CODE REVIEW

**Comment by StanleyDing** [comment]
Great job on the concise examples.

File: example/main.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/getamis/sirius/log"
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+const (
+	dkgProtocol = "/dkg/1.0.0"
+)
+
+func main() {
+	id := flag.Uint64("id", 0, "id for this node")
```
Comment (markya0616) [inline example/main.go:28]: put id into config?

File: example/config/id-3.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+rank: 0
+threshold:
+  dkg: 3
+  signer: 2
+  reshare: 3
+peers:
```
Comment (markya0616) [inline example/config/id-3.yaml:6]: may just pass peer id in, and we can get the address in our code?

File: example/node.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	maddr, err := multiaddr.NewMultiaddr(target)
+	if err != nil {
+		log.Warn("Cannot parse the target address", "target", target, "err", err)
+		return err
+	}
+
+	// Extract the peer ID from the multiaddr.
+	info, err := peer.AddrInfoFromP2pAddr(maddr)
+	if err != nil {
+		log.Error("Cannot parse addr", "addr", maddr, "err", err)
```
Comment (markya0616) [inline example/node.go:110]: Why log error here?

File: example/node.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		s.Reset()
+		return err
+	}
+
+	log.Info("Sent message", "peer", target)
+	return nil
+}
+
+// connect connects the host to the specified peer.
+func connect(host host.Host, target string) error {
```
Comment (markya0616) [inline example/node.go:99]: pass context in?

File: example/service.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	buf, err := ioutil.ReadAll(s)
+	if err != nil {
+		s.Reset()
+		log.Warn("Cannot read data from stream", "err", err)
+		return
+	}
+	s.Close()
+
+	// unmarshal it
+	proto.Unmarshal(buf, data)
```
Comment (markya0616) [inline example/service.go:66]: err = ?

File: example/service.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	}
+	s.dkg = d
+	return s, nil
+}
+
+func (p *service) Handle(s network.Stream) {
+	data := &dkg.Message{}
+	buf, err := ioutil.ReadAll(s)
+	if err != nil {
+		s.Reset()
```
Comment (markya0616) [inline example/service.go:59]: According the interface definition, 
```
// Reset closes both ends of the stream. Use this to tell the remote
// side to hang up and go away.
```
Do we want to disconnect if any error occur?

File: example/service.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+	// unmarshal it
+	proto.Unmarshal(buf, data)
+	if err != nil {
+		log.Error("Cannot unmarshal data", "err", err)
+		return
+	}
+
+	log.Info("Received request", "from", s.Conn().RemotePeer())
+	p.dkg.AddMessage(data)
```
Comment (markya0616) [inline example/service.go:73]: Check error?

File: example/node.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	basicHost, err := libp2p.New(context.Background(), opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return basicHost, nil
+}
+
+// send sends the proto message to specified peer.
+func send(host host.Host, target string, data proto.Message) error {
```
Comment (markya0616) [inline example/node.go:60]: pass context in

File: example/node.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	s, err := host.NewStream(context.Background(), info.ID, dkgProtocol)
+	if err != nil {
+		log.Warn("Cannot create a new stream", "from", host.ID(), "to", target, "err", err)
+		return err
+	}
+	writer := ggio.NewFullWriter(s)
+	err = writer.WriteMsg(data)
+	if err != nil {
+		log.Warn("Cannot write message to IO", "err", err)
+		s.Reset()
```
Comment (markya0616) [inline example/node.go:84]: Should we need to disconnect?

File: example/node.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+		return err
+	}
+	err = helpers.FullClose(s)
+	if err != nil {
+		log.Warn("Cannot close the stream", "err", err)
+		s.Reset()
```
Comment (markya0616) [inline example/node.go:90]: ditto

File: example/node.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
+// Copyright © 2020 AMIS Technologies
```
Comment (markya0616) [inline example/node.go:1]: we can move send() and connect() to pm.go? only peer manager needs these two methods.

File: example/utils.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	"fmt"
+
+	"github.com/getamis/alice/crypto/tss/dkg"
+	"github.com/getamis/sirius/log"
+)
+
+const (
+	typeDKG     int = 0
+	typeSigner  int = 1
+	typeReshare int = 2
```
Comment (markya0616) [inline example/utils.go:26]: rm typeReshare and typeSigner for now

File: example/service.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+func (p *service) Process() {
+	// 1. Start a DKG process.
+	p.dkg.Start()
+	defer p.dkg.Stop()
+
+	// 2. Connect the host to peers and send the peer message to them.
+	msg := p.dkg.GetPeerMessage()
+	peerIDs := getPeerIDs(p.pm.SelfID())
```
Comment (markya0616) [inline example/service.go:83]: get peer ids from peer manager?

File: example/utils.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+		resultType = "reshare"
+	}
+	return fmt.Sprintf("result/%s/%s.yaml", resultType, id)
+}
+
+func getID(id uint64) string {
+	return fmt.Sprintf("id-%d", id)
+}
+
+func getPeerIDs(selfID string) []string {
```
Comment (markya0616) [inline example/utils.go:65]: rm

File: example/utils.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/getamis/sirius/log"
+)
+
+const (
+	typeDKG     int = 0
+	typeSigner  int = 1
+	typeReshare int = 2
+)
+
+func writeDKGResult(id string, result *dkg.Result) error {
```
Comment (markya0616) [inline example/utils.go:29]: not check in the result files

File: example/README.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
+# TSS example
+
+This example demonstrates a simple p2p application using our TSS library. Let's assume we have 3 nodes where their ranks are all 0. These 3 nodes will interact with each other by using `go-libp2p` library. After each process (DKG, signer, and reshare), the results will be written in files located in `config/`.
+
+## Build
+```sh
+> cd example/
+> go build
```
Comment (markya0616) [inline example/README.md:8]: add into our Makefile

File: example/node.go (hunk 11 above)
Comment (bun919tw) [reply example/node.go:1]: I put these two functions in node.go because I think they are related to p2p.

File: example/utils.go (hunk 15 above)
Comment (bun919tw) [reply example/utils.go:29]: I don't get it lol

Comment (markya0616) [reply example/utils.go:27]: I mean you checked in the output files in this pull requests. They should not be checked in.

**Comment by yurenju** [comment]
looks good 😌

**Pull request merged**